docs/beefy.md

```diff
@@ -4,97 +4,399 @@ Related issues:
+3. BEEFY Justification for the latest mandatory block,
+
+Each message is additionally associated with a **topic**, which can be either:
+1. the round number (i.e. topic associated with a particular round),
+2. or the global topic (independent from the rounds).
+
+Round-specific topic should only be used to gossip the votes, other messages are gossiped
+periodically on the global topic. Let's now dive into description of the messages.
+
+- **Votes**
+  - Votes are sent on the round-specific topic.
+  - Vote is considered valid when:
+    - The commitment matches local commitment.
+    - The validator is part of the current validator set.
+    - The signature is correct.
+
+- **BEEFY Justification**
+  - Justifications are sent on the global topic.
+  - Justification is considered worthwhile to gossip when:
+    - It is for a recent (implementation specific) round or the latest mandatory round.
+    - All signatures are valid and there is at least `2/3rd + 1` of them.
+    - Signatorees are part of the current validator set.
+  - Mandatory justifications should be announced periodically (see also [Lean BEEFY](#2-lean-beefy)).
+
+## Misbehavior
+
+Similarily to other PoS protocols, BEEFY considers casting two different votes in the same round a
+misbehavior. I.e. for a particular `round_number`, the validator produces signatures for 2 different
+`Commitment`s and broadcasts them. This is called **equivocation**.
+
+On top of this, voting on an incorrect **payload** is considered a misbehavior as well, and since
+we piggy-back on GRANDPA there is no ambiguity in terms of the fork validators should be voting for.
+
+Misbehavior should be penalized. If more validators misbehave in the exact same `round` the penalty
+should be more severe, up to the entire bonded stake in case we reach `1/3rd + 1` validators
+misbehaving.
+
+<details>
+<summary>Open questions</summary>
+
+// TODO [ToDr] Penalization formula.
+// TODO [ToDr] If there are misbehaving validators in GRANDPA should we penalize them twice?
+IMHO yes, cause they should be relying on GRANDPA finality proof, not just the votes.
+
+</details>
+
+# Implementation
+
+TODO
+
+- What if grandpa < beefy?
+- Missing mandatory blocks
+- Extra assumptions and consequences (session start block selection) [Lean BEEFY]
+
+///
+When we start the worker, we should:
+1. Ask the Client for the Best BEEFY block
+   (TODO: what if we don't have that API?)
+2. Traverse up to "Session Length" number of blocks forward to find a session transition.
+3. Figure out if we need to vote on the mandatory block of the next session
+   (i.e. we are doing BEEFY Catch-up)
+4. or if we can follow the power-of-two rule.
+
+With LEAN BEEFY:
+1. Start with Best GRANDPA block
+2. Traverse up to "Session Length" number of blocks backwards to find the session start block.
+3. ASSUME this block is BEEFY-finalized, even though we haven't seen BEEFY Justification.
+
+
+//TODO [ToDr] How to figure out the `session_start` block?
```

Comment:
PR https://github.com/paritytech/substrate/pull/10669 implement `current_session_start_block_num` storage for session pallet. That should work fine, right?